### PR TITLE
issue #106: improve exception handling in `fertiscan`

### PR DESCRIPTION
**Pull request by k-allagbe**
Changes:

- For each of metadata and queries, there now is a module for all the errors.
- The errors now have a hierarchy which helps when handling a group of exceptions at once

```
Query exceptions hierarchy:

    Exception
    |__QueryError
       |__InspectionQueryError
       |__LabelInformationQueryError
       |__LabelDimensionQueryError
       |__MetricQueryError
       |__UnitQueryError
       |__ElementCompoundQueryError
       |__MicronutrientQueryError
       |__GuaranteedAnalysisQueryError
       |__OrganizationQueryError
       |__OrganizationInformationQueryError
       |__LocationQueryError
       |__RegionQueryError
       |__ProvinceQueryError
       |__SpecificationQueryError
       |__SubLabelQueryError
       |__SubTypeQueryError

Each QueryError sub type exception has specific errors related to creation, retrieval, updating, and deletion, as well as a 'not found' error.
```

```
Metadata exceptions hierarchy:

    Exception
    |__MetadataError
       |__BuildInspectionImportError
       |__BuildInspectionExportError
       |__NPKError
```

- Because the error handling pattern is the same for pretty much all the query functions, we use a decorator to minimize code repetition:
```python
def handle_query_errors(error_cls=QueryError):
    """Decorator for handling query errors."""

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except QueryError:
                raise
            except Error as db_error:
                raise error_cls(f"Database error: {db_error}") from db_error
            except Exception as e:
                raise error_cls(f"Unexpected error: {e}") from e

        return wrapper

    return decorator

```

The code using a decorator:

```python
@handle_query_errors(LabelInformationRetrievalError)
def get_label_information_json(cursor, label_info_id) -> dict:
    query = """
        SELECT get_label_info_json(%s);
        """
    cursor.execute(query, (str(label_info_id),))
    label_info = cursor.fetchone()
    if label_info is None or label_info[0] is None:
        raise LabelInformationNotFoundError(
            "Error: could not get the label information: " + str(label_info_id)
        )
    return label_info[0]
```

is equivalent to:

```python
def get_label_information_json(cursor, label_info_id) -> dict:
    try:
        query = """
            SELECT get_label_info_json(%s);
            """
        cursor.execute(query, (str(label_info_id),))
        label_info = cursor.fetchone()
        if label_info is None or label_info[0] is None:
            raise LabelInformationNotFoundError(
                "Error: could not get the label information: " + str(label_info_id)
            )
        return label_info[0]
    except QueryError:
        raise
    except Error as db_error:
        raise LabelInformationRetrievalError(f"Database error: {db_error}") from db_error
    except Exception as e:
        raise LabelInformationRetrievalError(f"Unexpected error: {e}") from e
```

Note that `LabelInformationNotFoundError` is a `QueryError`.

### CODE REVIEW

File: fertiscan/db/metadata/errors.py
Comment (Francois-Werbrouck) [inline fertiscan/db/metadata/errors.py:1]: Why are we switching to custom error files instead of staying with our modules errors?

File: fertiscan/db/queries/label/__init__.py [hunk 1]
```diff
@@ -84,35 +93,36 @@ def get_label_information(cursor, label_information_id):
...
+        FROM 
+            label_information
+        WHERE 
+            id = %s
+        """
+    cursor.execute(query, (label_information_id,))
+    return cursor.fetchone()
+
+
+@handle_query_errors(LabelInformationRetrievalError)
```
Comment (Francois-Werbrouck) [inline fertiscan/db/queries/label/__init__.py:125]: I dont understand here we use `LabelInformationRetrievalError` but we raise `LabelInformationNotFoundError` (1/2)

File: fertiscan/db/queries/label/__init__.py [hunk 2]
```diff
@@ -124,23 +134,19 @@ def get_label_information_json(cursor, label_info_id) -> dict:
...
+    cursor.execute(query, (str(label_info_id),))
+    label_info = cursor.fetchone()
+    if label_info is None or label_info[0] is None:
+        raise LabelInformationNotFoundError(
+            "Error: could not get the label information: " + str(label_info_id)
+        )
+    return label_info[0]
 
 
+@handle_query_errors(LabelInformationQueryError)
```
Comment (Francois-Werbrouck) [inline fertiscan/db/queries/label/__init__.py:149]: and here we use `LabelInformationQueryError` but raise `LabelInformationNotFoundError`. Why is it in one case a query error and the other a retrival error. Also what is the difference between a retrival error and NotFoundError. I like how the handle_querry_error() works, but I dont get how you use it in this case. (2/2)

File: fertiscan_pyproject.toml [hunk 3]
```diff
@@ -4,7 +4,7 @@ build-backend = "setuptools.build_meta"
 
 [project]
 name = "fertiscan_datastore"
-version = "1.0.3"
+version = "1.0.4"
 authors = [
   { name="Francois Werbrouck", email="francois.werbrouck@inspection.gc.ca" }
```
Comment (Francois-Werbrouck) [inline fertiscan_pyproject.toml:9]:
```suggestion
  { name="Francois Werbrouck", email="francois.werbrouck@inspection.gc.ca" },
```

File: fertiscan_pyproject.toml [hunk 4]
```diff
@@ -4,7 +4,7 @@ build-backend = "setuptools.build_meta"
...
   { name="Kotchikpa Guy-Landry Allagbe" , email = "kotchikpaguy-landry.allagbe@inspection.gc.ca"}
```
Comment (Francois-Werbrouck) [inline fertiscan_pyproject.toml:10]: I think there might be an error. Check suggestion above

File: fertiscan/db/metadata/errors.py
Comment (k-allagbe) [reply fertiscan/db/metadata/errors.py:1]: Just a common pattern I've seen from many other libraries and copied.
- https://github.com/celery/celery/blob/main/celery/exceptions.py
- https://github.com/fastapi/fastapi/blob/master/fastapi/exceptions.py
- https://github.com/psf/requests/blob/main/src/requests/exceptions.py

Personally, it's easier to maintain for me that way.

File: fertiscan/db/queries/label/__init__.py (hunk 1 above)
Comment (k-allagbe) [reply fertiscan/db/queries/label/__init__.py:125]: Yes. It only means that both could be raised by the same function: LabelInformationNotFoundError in the particular case where None is returned and LabelInformationRetrievalError in the other cases.

File: fertiscan/db/queries/label/__init__.py (hunk 2 above)
Comment (k-allagbe) [reply fertiscan/db/queries/label/__init__.py:149]: Label dimension is not really a LabelInformation. Using LabelInformationRetrievalError did not seem appropriate, hence, I use LabelInformationQueryError which is the "default" exception for LabelInformations. But if this does not seem natural, please suggest an alternative.

Note that I now raise LabelInformationQueryError as well when no data is returned.

File: fertiscan_pyproject.toml (hunk 4 above)
Comment (k-allagbe) [reply fertiscan_pyproject.toml:10]: Wonder how it passed workflows without failing.

File: fertiscan/db/metadata/errors.py
Comment (Francois-Werbrouck) [reply fertiscan/db/metadata/errors.py:1]: We made the decision in the past to not use error file and make module custom exceptions. Both are commonly used, both work, but making the switch out of the blue is a bit unexpected.

I personally dont think it is easier to maintain, I think both have their fair share of advantages in their usage.

File: fertiscan/db/queries/label/__init__.py (hunk 2 above)
Comment (Francois-Werbrouck) [reply fertiscan/db/queries/label/__init__.py:149]: So from my understanding:
If when retrieving a label and no data is return (result = None )we get `LabelInformationNotFoundError` which raise `LabelInformationRetrievalError` and we also trigger `LabelInformationQueryError`  ?
If there is a random error during the retrieval of a label or label_dimension we get `LabelInformationQueryError`
If when retrieving a label_dimension and no data is return (result = None ) we only get `LabelInformationQueryError`? 

I understand that a label_dimension is not the same as a label_information. (Its an abstract concept representing the data of the real thing) I think this issue has a goal of giving more insight on what is happening. **I'd give it a retrieval or not found type of error to help the user debug**(maybe even a new exception).

 Side note, the function is mostly used in testing and if a label_information doesn't exist, it's dimension should not exists.

File: fertiscan_pyproject.toml (hunk 4 above)
Comment (Francois-Werbrouck) [reply fertiscan_pyproject.toml:10]: same here. tagging @SonOfLope so he can be noticed of this weird behavior

Comment (SonOfLope) [reply fertiscan_pyproject.toml:10]: Thats because it compares with the releases and latest  release was 1.0.2. The errors is that release 1.0.3 doesnt exist. Thats because it shouldve been created when this PR was merged : 
https://github.com/ai-cfia/ailab-datastore/pull/178
Problem is, releases are done when changes are made to folder fertiscan or nachet or datastore but that PR only changed the pyproject and the test folder. Nice edge case guys ! I will update pipeline to handle this case :)

File: fertiscan/db/metadata/errors.py
Comment (k-allagbe) [reply fertiscan/db/metadata/errors.py:1]: I do not think both are commonly used, especially in libraries where clear separation of concerns is important. Although, by that same argument, having all the exception categories in the same module is also not very appropriate. The most rigorous approach would be to have an error file for each category, but I think we can agree that it's polluting. I think the current approach is a good balance between good concern separation and tidiness.

File: fertiscan/db/queries/label/__init__.py (hunk 2 above)
Comment (k-allagbe) [reply fertiscan/db/queries/label/__init__.py]: I created new errors for label dimension. Seems neater.

Note: I've updated the description of the PR to explain what's going on.

File: fertiscan/db/metadata/errors.py
Comment (k-allagbe) [reply fertiscan/db/metadata/errors.py:1]: Haha, mispoke. Meant the single error file.

File: fertiscan/db/queries/errors.py [hunk 5]
```diff
@@ -0,0 +1,606 @@
+"""
+Query exceptions hierarchy:
+
+    Exception
+    |__QueryError
+       |__InspectionQueryError
```
Comment (Francois-Werbrouck) [inline fertiscan/db/queries/errors.py:6]: Could we just add another tier for each object? I think it would helps classifying the errors? WDYT?

**Review by Francois-Werbrouck — APPROVED** [review]
After a fresh review, I think the error file will help maintain the project in the long run!

File: fertiscan/db/queries/errors.py (hunk 5 above)
Comment (k-allagbe) [reply fertiscan/db/queries/errors.py:6]: How do you mean?

Comment (Francois-Werbrouck) [reply fertiscan/db/queries/errors.py:6]:
```Python
"""
Exception
    |__QueryError
       |__Inspection
          |__InspectionQueryError
          |__InspectionXXXXXError
       |__Label_Information
          |__Label_Specific_errors
       |__Metrics
          |__...
"""
```

Comment (Francois-Werbrouck) [reply fertiscan/db/queries/errors.py:6]: I feel like having all the Query errors under the umbrella of querry error is not specific enough. 
Classifying the errors in smaller groups within the Query Error type seems to be more instinctive to me (having errors per objects)

Comment (k-allagbe) [reply fertiscan/db/queries/errors.py:6]: Please note that the actual hierarchy is:
```
    Exception
    |__QueryError
       |__InspectionQueryError
          |__InspectionNotFoundError
          |__InspectionCreationError
          |__...
...
```

Hence the clarification:
```
Each QueryError sub type exception has specific errors related to creation, retrieval, updating, and deletion, as well as a 'not found' error.
```

Comment (Francois-Werbrouck) [reply fertiscan/db/queries/errors.py:6]: Thank you for the clarification. From my understanding of the file Comment it was not the case.

**Comment by k-allagbe** [comment]
This issue will not fix #108 as that is code `datastore` code that might affect nachet.

File: fertiscan/db/queries/errors.py (hunk 5 above)
Comment (k-allagbe) [reply fertiscan/db/queries/errors.py:6]: I updated the note.
